src/jit/stacklevelsetter.cpp

```diff
@@ -0,0 +1,264 @@
+    , currentStackLevel(0)
+    , maxStackLevel(0)
+#if !FEATURE_FIXED_OUT_ARGS
+    , framePointerRequired(compiler->codeGen->isFramePointerRequired())
+#endif // !FEATURE_FIXED_OUT_ARGS
+    , memAllocator(compiler, CMK_fgArgInfoPtrArr)
+    , putArgNumSlots(&memAllocator)
+{
+}
+
+//------------------------------------------------------------------------
+// DoPhase: Calculate stack slots numbers for outgoing args.
+//
+// Notes:
+//   For non-x86 platforms it calculates the max number of slots
+//   that calls inside this method can push on the stack.
+//   This value is used for sanity checks in the emitter.
+//
+//   Stack slots are pointer-sized: 4 bytes for x86, 8 bytes for amd64.
+//
+//   For x86 it also sets throw-helper blocks incoming stack depth and set
+//   framePointerRequired when it is necessary. These values are used to pop
+//   pushed args when an exception occurs.
+void StackLevelSetter::DoPhase()
+{
+    for (BasicBlock* block = comp->fgFirstBB; block != nullptr; block = block->bbNext)
+    {
+        ProcessBlock(block);
+    }
+#if !FEATURE_FIXED_OUT_ARGS
+
+    if (framePointerRequired && !comp->codeGen->isFramePointerRequired())
+    {
+        JITDUMP("framePointerRequired is not set when it is required\n");
+        comp->codeGen->resetWritePhaseForFramePointerRequired();
+        comp->codeGen->setFramePointerRequired(true);
+    }
+#endif // !FEATURE_FIXED_OUT_ARGS
+    assert(maxStackLevel <= comp->fgGetPtrArgCntMax());
+    if (maxStackLevel != comp->fgGetPtrArgCntMax())
+    {
+        JITDUMP("fgPtrArgCntMax was calculated wrong during the morph, the old value: %u, the right value: %u.\n",
+                comp->fgGetPtrArgCntMax(), maxStackLevel);
+        comp->fgSetPtrArgCntMax(maxStackLevel);
+    }
+}
+
+//------------------------------------------------------------------------
+// ProcessBlock: Do stack level calculations for one block.
+//
+// Notes:
+//   Block starts and ends with an empty outgoing stack.
+//   Nodes in blocks are iterated in the reverse order to memorize GT_PUTARG_STK
+//   and GT_PUTARG_SPLIT stack sizes.
+//
+// Arguments:
+//   block - the block to process.
+//
+void StackLevelSetter::ProcessBlock(BasicBlock* block)
+{
+    assert(currentStackLevel == 0);
+    LIR::ReadOnlyRange& range = LIR::AsRange(block);
+    for (auto i = range.rbegin(); i != range.rend(); ++i)
+    {
+        GenTree* node = *i;
+        if (node->OperIsPutArgStkOrSplit())
+        {
+            GenTreePutArgStk* putArg   = static_cast<GenTreePutArgStk*>(node);
+            unsigned          numSlots = putArgNumSlots[putArg];
+            putArgNumSlots.Remove(putArg);
+            SubStackLevel(numSlots);
+        }
+
+#if !FEATURE_FIXED_OUT_ARGS
+        // Set throw blocks incoming stack depth for x86.
+        bool operMightThrow = ((node->gtFlags & GTF_EXCEPT) != 0);
+        if (!framePointerRequired && comp->fgUseThrowHelperBlocks() && operMightThrow)
```

Comment:
Seems like you should put `operMightThrow` first, since it's a local, to potentially avoid the other checks.